Exclude upsample_bilinear2d.vec from default core ATen decomposition table

GregoryComer opened the pull request:
As upsample_bilinear2d.vec is a core ATen op, it should not be decomposed by default in the export path. Because the operator has CompositeImplicitAutograd dispatch, its decomposition is registered by default. This change adds an override list for CIA decompositions being registered in the default decomp table.
In the long-term, we likely will want to exclude decompositions for all core-tagged CIA ops, but this will require all consumers to be ready to handle the remaining three ops: upsample_nearest2d.vec, avg_pool1d, and adaptive_avg_pool1d. Until they are ready, I believe an explicit override list is the safest option.

Additionally, I've also removed the ExecuTorch XNNPACK delegate ConvertToUpsampleBilinear2d pass, as the pass breaks (and is not needed), given that the op is not decomposed. The purpose of this pass was originally to pattern match the decomposition and un-decomposite it, but this is no longer necessary.

GregoryComer commented on test/export/test_export.py:
```diff
@@ -10092,30 +10092,30 @@ def forward(self, x):
         )
 
     def test_preserve_cia_op(self):
```
I've updated this test to use Trilinear3d to preserve the original intent of the test. Since Bilinear2d is core, it is always preserved (and also the test fails, given that it can't pop the decomp). Trilinear3d is not core and is composite implicit autograd.

bdhirsh commented on torch/export/decomp_utils.py:
```diff
@@ -13,6 +13,13 @@
 __all__ = ["CustomDecompTable"]
 
 
+PRESERVED_ATEN_CIA_OPS = set(
```
@tugsbayasgalan it seems to me like we shouldn't need to do this now that your CIA refactors have landed?

`upsample_bilinear2d.vec` should be preserved automatically unless the user specifies it in their decomp table

bdhirsh commented on torchgen/native_function_generation.py:
```diff
@@ -405,14 +406,14 @@ def add_generated_native_functions(
                 f.has_composite_implicit_autograd_kernel for f in d.values()
             )
-            if are_manual or has_view_ops or are_composite_implicit:
+            if are_manual or has_view_ops or are_composite_implicit and not is_core:
```
hmm @GregoryComer one invariant I think we want to keep is that for **any** ATen ops that we auto-generate here, they have an `autogen: <op_name>` entry in `native_functions.yaml` (e.g. here: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/native_functions.yaml#L12610)

It looks like that empirically isn't completely true today... given that I see `autogen: upsample_bilinear2d.vec_out` in the yaml, but `torch.ops.aten.upsample_bilinear2d.vec_out` doesn't exist (@larryliu0820 does this ring a bell for you?)

But at the very least, we can keep the invariant in the other direction: instead of **always** doing this for ops marked as "core", can we also require them to exist in the yaml and have the `autogen` tag?

GregoryComer replied:
Thanks. I have a separate PR out to allow autogen of core CIA ops - right now, autogen is skipped for all CIA ops. Any concerns with this approach: https://github.com/pytorch/pytorch/pull/141797?

GregoryComer replied:
FYI This PR (https://github.com/pytorch/pytorch/pull/1417970) is included here. I was hitting issues with ghstack due to the cross-repo dep, but I can rebase if/once that lands.

Regarding this check, I see there is a needs_out check already in place that checks to see if there is an autogen variant. However, it is getting skipped for CIA ops. I could just remove the are_composite_implicit from above, if we want to do that.

larryliu0820 replied:
@bdhirsh We will continue/skip if a native function has a CIA kernel in line 417, and only check the `autogen` tag in line 471. So basically "having CIA kernel" overrides "having `autogen` tag"

bdhirsh commented on torch/export/decomp_utils.py:
```diff
@@ -38,7 +45,8 @@ def __init__(self):
 
         for op in _collect_all_valid_cia_ops_for_aten_namespace():
-            self.decomp_table[op] = _get_decomp_for_cia(op)
+            if op not in PRESERVED_ATEN_CIA_OPS:
```
oh I think i see. Instead of making the change here, can we:

(1) use `torch._decomp.core_aten_decompositions()`

(2) tweak that function [here](https://github.com/pytorch/pytorch/blob/main/torch/_decomp/__init__.py#L298) - today, it looks like it unconditionally adds all CIA decomps to the list of decomps. But instead, we can filter it to not add the decomp to the list if the op has a "core" tag. Something like:
```
    for op in _collect_all_valid_cia_ops_for_aten_namespace():
        if torch.Tag.core not in op.tags:
            decomp_table[op] = _get_decomp_for_cia(op)
        
```

tugsbayasgalan replied:
Ah sorry the PR that makes core_aten_decomp as alias to export table is still not landed yet.... 
But i think we should do (2) yeah. Maintaining another list PRESERVED_ATEN_CIA_OPS makes the current decomp logic even more complicated.

tugsbayasgalan replied:
Ahh i see so you want to only do this for this op for this PR... Could you just hide this logic here https://www.internalfb.com/code/fbsource/[2f8dd11dd8963caf6e9f4ce48464ae111b55f1ad]/fbcode/caffe2/torch/_export/utils.py?lines=1072 Basically this is the only util function we use to query all valid cia ops we can mess qround. So we exclude it from this list, export system will automatically think this op is not a CIA i think...

tugsbayasgalan replied:
Ping here!

GregoryComer replied:
Talked offline with Tugsuu and he was able to explain more of the context here. Shouldn't be an issue for this change.

GregoryComer committed:
Exclude upsample_bilinear2d.vec and nearest2d.vec from default export decomposition table (#141791)

Summary:
As upsample_bilinear2d.vec and upsample_nearest2d.vec are core ATen ops, they should not be decomposed by default in the export path. Because the operators have CompositeImplicitAutograd dispatch, their decomposition is registered by default. This change adds an override list for CIA decompositions being registered in the default decomp table.

In the long-term, we likely will want to exclude decompositions for all core-tagged CIA ops, but this will require all consumers to be ready to handle the remaining two ops, avg_pool1d, and adaptive_avg_pool1d. Until they are ready, I believe an explicit override list is the safest option.

Additionally, I've also removed the ExecuTorch XNNPACK delegate ConvertToUpsampleBilinear2d pass, as the pass breaks (and is not needed), given that the op is not decomposed. The purpose of this pass was originally to pattern match the decomposition and recompose it, but this is no longer necessary.

X-link: https://github.com/pytorch/executorch/pull/7126


Test Plan:
Added a new test (`test_default_decomposition_core_cia_ops`) in test_export.py to verify that upsample_bilinear2d.vec (and in the future, other core-tagged CIA ops) are not decomposed by default. Also, I manually validated end to end with ExecuTorch that the op is not decomposed in to_edge (see N6238522).

```
buck test //caffe2/test:test_export -- test_default_decomposition_core_cia_ops
```

Reviewed By: digantdesai

Differential Revision: D66575454

huydhn commented:
@pytorchbot merge
